core/src/main/java/com/sequenceiq/cloudbreak/service/multiaz/MultiAzCalculatorService.java

```diff
@@ -248,4 +254,123 @@ private boolean isPreferredSubnetsSpecifiedForScaling(NetworkScaleDetails networ
+                if (isNullOrEmpty(instanceMetaData.getSubnetId())) {
+                    GroupPlacement groupPlacement = getGroupPlacementForInstanceGroupName(instanceMetaData.getInstanceGroupName(),
+                            instanceMetaData.getInstanceGroup().getHints(), groupPlacementList);
+                    Integer numberOfInstanceInASubnet = searchTheSmallestInstanceCountForUsage(groupPlacement.getSubnetUsage());
+                    String leastUsedSubnetId = searchTheSmallestUsedID(groupPlacement.getSubnetUsage(), numberOfInstanceInASubnet);
+
+                    LOGGER.info("{} instance (Group {}) placed at {}", instanceMetaData.getInstanceGroupName(), groupPlacement.getName(),
+                            groupPlacement.getAvailabilityZone());
+                    instanceMetaData.setSubnetId(leastUsedSubnetId);
+                    instanceMetaData.setAvailabilityZone(groupPlacement.getAvailabilityZone());
+
+                    groupPlacement.addInstanceGroupName(instanceMetaData.getInstanceGroupName());
+                    groupPlacement.increaseSubnetUsage(leastUsedSubnetId);
+                    subnetIds.add(leastUsedSubnetId);
+                    availabilityZones.add(groupPlacement.getAvailabilityZone());
+                }
+            }
+            updateInstanceGroup(instanceGroup, subnetIds, availabilityZones);
+        }
+    }
+
+    private void updateInstanceGroup(InstanceGroup instanceGroup, List<String> subnetIds, Set<String> azNames) {
+        Map<String, Object> networkAttributes = Optional
+                .of(instanceGroup.getInstanceGroupNetwork())
+                .map(InstanceGroupNetwork::getAttributes)
+                .map(Json::getMap)
+                .orElse(new HashMap<>());
+        networkAttributes.put(SUBNET_IDS, subnetIds);
+        instanceGroup.getInstanceGroupNetwork().setAttributes(new Json(networkAttributes));
+
+        Set<AvailabilityZone> availabilityZones = new HashSet<>();
+        azNames.forEach(az -> {
+            AvailabilityZone availabilityZone = new AvailabilityZone();
+            availabilityZone.setAvailabilityZone(az);
+            availabilityZone.setInstanceGroup(instanceGroup);
+            availabilityZones.add(availabilityZone);
+
+        });
+        instanceGroup.setAvailabilityZones(availabilityZones);
+    }
+
+    private GroupPlacement getGroupPlacementForInstanceGroupName(String instanceGroupName, Set<String> hints, Set<GroupPlacement> groupPlacements) {
+        if (hints == null || hints.isEmpty()) {
+            return groupPlacements
+                    .stream()
+                    .findFirst()
+                    .get();
+        }
+
+        return groupPlacements
+                .stream()
+                .filter(gp -> hints.contains(gp.getName()))
+                .filter(gp -> !gp.containsInstanceGroupName(instanceGroupName))
```

Comment:
should we have it because it means only one node for an instance group can be part of an AZ which does not seem right. If number of nodes in instance group is more than hints then we should map multiple nodes to a given hint i.e. put them in same AZ.